src/iceberg/manifest/manifest_reader.cc

```diff
@@ -19,61 +19,867 @@
+  } else {
+    ICEBERG_ASSIGN_OR_RAISE(
+        projected_data_file_schema,
+        ProjectSchema(std::move(data_file_schema), columns_, case_sensitive_));
+  }
+
+  ICEBERG_RETURN_UNEXPECTED(OpenReader(std::move(projected_data_file_schema)));
+  ICEBERG_DCHECK(file_reader_ != nullptr, "File reader should be initialized");
+
+  std::vector<ManifestEntry> manifest_entries;
+  ICEBERG_ASSIGN_OR_RAISE(auto arrow_schema, file_reader_->Schema());
+  internal::ArrowSchemaGuard schema_guard(&arrow_schema);
+
+  // Get evaluators if needed
+  Evaluator* evaluator = nullptr;
+  InclusiveMetricsEvaluator* metrics_evaluator = nullptr;
+  if (HasPartitionFilter() || HasRowFilter()) {
+    ICEBERG_ASSIGN_OR_RAISE(evaluator, GetEvaluator());
+  }
+  if (HasRowFilter()) {
+    ICEBERG_ASSIGN_OR_RAISE(metrics_evaluator, GetMetricsEvaluator());
+  }
+
+  while (true) {
+    ICEBERG_ASSIGN_OR_RAISE(auto result, file_reader_->Next());
+    if (!result.has_value()) {
+      break;  // EOF
+    }
+
+    internal::ArrowArrayGuard array_guard(&result.value());
+    ICEBERG_ASSIGN_OR_RAISE(
+        auto entries,
+        ParseManifestEntry(&arrow_schema, &result.value(), *file_schema_, first_row_id_));
+
+    for (auto& entry : entries) {
+      ICEBERG_RETURN_UNEXPECTED(inheritable_metadata_->Apply(entry));
+
+      if (only_live && !entry.IsAlive()) {
+        continue;
+      }
+
+      if (needs_filtering) {
+        ICEBERG_DCHECK(entry.data_file != nullptr, "Data file cannot be null");
+        if (evaluator) {
+          ICEBERG_ASSIGN_OR_RAISE(bool partition_match,
+                                  evaluator->Evaluate(entry.data_file->partition));
+          if (!partition_match) {
+            continue;
+          }
+        }
+        if (metrics_evaluator) {
+          ICEBERG_ASSIGN_OR_RAISE(bool metrics_match,
+                                  metrics_evaluator->Evaluate(*entry.data_file));
+          if (!metrics_match) {
+            continue;
+          }
+        }
+        if (!InPartitionSet(*entry.data_file)) {
+          continue;
+        }
+      }
+
+      manifest_entries.push_back(std::move(entry));
+    }
+  }
+
+  return manifest_entries;
+}
+
+Result<std::vector<ManifestFile>> ManifestListReaderImpl::Files() const {
+  std::vector<ManifestFile> manifest_files;
+  ICEBERG_ASSIGN_OR_RAISE(auto arrow_schema, reader_->Schema());
+  internal::ArrowSchemaGuard schema_guard(&arrow_schema);
+  while (true) {
+    ICEBERG_ASSIGN_OR_RAISE(auto result, reader_->Next());
+    if (result.has_value()) {
+      internal::ArrowArrayGuard array_guard(&result.value());
+      ICEBERG_ASSIGN_OR_RAISE(
+          auto parse_result, ParseManifestList(&arrow_schema, &result.value(), *schema_));
+      manifest_files.insert(manifest_files.end(),
+                            std::make_move_iterator(parse_result.begin()),
+                            std::make_move_iterator(parse_result.end()));
+    } else {
+      // eof
+      break;
+    }
```

Comment:
```suggestion
    if (!result.has_value()) {
      // eof
      break;
    }
    internal::ArrowArrayGuard array_guard(&result.value());
    ICEBERG_ASSIGN_OR_RAISE(
        auto parse_result, ParseManifestList(&arrow_schema, &result.value(), *schema_));
    manifest_files.insert(manifest_files.end(),
                          std::make_move_iterator(parse_result.begin()),
                          std::make_move_iterator(parse_result.end()));
```
I prefer to write this way, with one less tab.